torch/autograd/gradcheck.py

```diff
@@ -279,6 +360,93 @@ def vjp(v):
         return fail_test(get_failed_batched_grad_test_msg(output_idx, input_idx, res, exp))
 
 
+def test_backward_mul_by_grad_output(fail_test, func, inputs, check_sparse_nnz) -> bool:
+    # Tests that backward is multiplied by grad_output
+    output = _differentiable_outputs(func(*inputs))
+    diff_input_list: List[torch.Tensor] = list(iter_tensors(inputs, True))
+    if not diff_input_list:
+        raise RuntimeError("no Tensors requiring grad found in input")
+    grads_input = torch.autograd.grad(output, diff_input_list,
+                                      [torch.zeros_like(o, memory_format=torch.legacy_contiguous_format) for o in output],
+                                      allow_unused=True)
+    for gi, di in zip(grads_input, diff_input_list):
+        if gi is None:
+            continue
+        if isinstance(gi, torch.Tensor) and gi.layout != torch.strided:
+            if gi.layout != di.layout:
+                return fail_test('grad is incorrect layout (' + str(gi.layout) + ' is not ' + str(di.layout) + ')')
+            if gi.layout == torch.sparse_coo:
+                if gi.sparse_dim() != di.sparse_dim():
+                    return fail_test('grad is sparse tensor, but has incorrect sparse_dim')
+                if gi.dense_dim() != di.dense_dim():
+                    return fail_test('grad is sparse tensor, but has incorrect dense_dim')
+            gi = gi.to_dense()
+            di = di.to_dense()
+
+        if check_sparse_nnz:
+            if not torch.allclose(gi, torch.zeros_like(gi)):
+                return fail_test('backward not multiplied by grad_output')
+        elif not gi.eq(0).all():
+            return fail_test('backward not multiplied by grad_output')
+        if gi.dtype != di.dtype or gi.device != di.device or gi.is_sparse != di.is_sparse:
+            return fail_test("grad is incorrect type")
+        if gi.size() != di.size():
+            return fail_test('grad is incorrect size')
+    return True
+
+
+def test_undefined_grad(fail_test, func, inputs) -> bool:
+    output = _differentiable_outputs(func(*inputs))
```

Comment:
Previously, we were computing `func(*input)` once (see previous like 498). Now, we're doing it twice, once in test_undefined_grad and once in test_backward_mul_by_grad_output.

I assume this is fine w.r.t performance? The bulk of the work in gradcheck is probably the numeric jacobian computation.

Reply:
I think it might be because can't backward through the same graph multiple times (when there are saved variables that are needed for backward).

Reply:
This is a good catch, and it should be fixed now. Numeric jacobian is indeed the bulk of the computation right now, but only because its computing func so many times. Currently, we call func at least 2 * num_outputs * num_inputs * dim_inputs  times to compute numerical jacobian (there might be a way to get it to 2 * num_outputs * dim_inputs, but requires more extensive refactoring). So, though it might seem like a relatively small improvement right now, I think this could be significant especially once we use it in the fast-gradcheck version, which only needs to call func 2 * num_inputs times to compute numerical jacobian.